src/io/platform/f3xx/f302x8/I2Cf302x8.cpp

```diff
@@ -91,57 +91,99 @@ I2Cf302x8::I2Cf302x8(Pin sclPin, Pin sdaPin) : I2C(sclPin, sdaPin) {
+                                                       bytes, length,
+                                                       DEFAULT_I2C_TIMEOUT);
+    return halToI2CStatus(status);
 }
 
 /**
  * Override the default multi byte read since the F302x8 supports doing
  * this via the HAL.
  */
-void I2Cf302x8::read(uint8_t addr, uint8_t* bytes, uint8_t length) {
-    HAL_I2C_Master_Receive(&halI2C, addr << 1, bytes, length,
-                           DEFAULT_I2C_TIMEOUT);
+I2C::I2CStatus I2Cf302x8::read(uint8_t addr, uint8_t* bytes, uint8_t length) {
+    HAL_StatusTypeDef status = HAL_I2C_Master_Receive(&halI2C, addr << 1,
+                                                      bytes, length,
+                                                      DEFAULT_I2C_TIMEOUT);
+    return halToI2CStatus(status);
 }
 
-void I2Cf302x8::writeMemReg(uint8_t addr, uint32_t memAddress, uint8_t byte, uint16_t memAddSize, uint8_t maxWriteTime) {
+I2C::I2CStatus I2Cf302x8::writeMemReg(uint8_t addr, uint32_t memAddress,
+                                      uint8_t byte, uint16_t memAddSize,
+                                      uint8_t maxWriteTime) {
     uint16_t memAddress16 = memAddress;
-    HAL_I2C_Mem_Write(&halI2C, addr << 1, memAddress16, memAddSize, &byte, 1, DEFAULT_I2C_TIMEOUT);
-    HAL_Delay(maxWriteTime);// TODO: If possible, replace with a flag check instead of hardcoded timer
+    HAL_StatusTypeDef status = HAL_I2C_Mem_Write(&halI2C, addr << 1,
+                                                 memAddress16, memAddSize,
+                                                 &byte, 1, DEFAULT_I2C_TIMEOUT);
+    return halToI2CStatus(status);
 }
 
-uint8_t I2Cf302x8::readMemReg(uint8_t addr, uint32_t memAddress, uint8_t* byte, uint16_t memAddSize) {
+I2C::I2CStatus I2Cf302x8::readMemReg(uint8_t addr, uint32_t memAddress,
+                                     uint8_t* byte,
+                                     uint16_t memAddSize) {
     uint16_t memAddress16 = memAddress;
-    return HAL_I2C_Mem_Read(&halI2C, addr << 1, memAddress16, memAddSize, byte, 1, DEFAULT_I2C_TIMEOUT);
+    HAL_StatusTypeDef status = HAL_I2C_Mem_Read(&halI2C, addr << 1,
+                                                memAddress16, memAddSize, byte,
+                                                1, DEFAULT_I2C_TIMEOUT);
+    return halToI2CStatus(status);
 }
 
-void I2Cf302x8::writeMemReg(uint8_t addr, uint32_t memAddress, uint8_t bytes[], uint8_t size, uint16_t memAddSize,
-                            uint8_t maxWriteTime) {
+I2C::I2CStatus I2Cf302x8::writeMemReg(uint8_t addr, uint32_t memAddress,
+                                      uint8_t* bytes, uint8_t size,
+                                      uint16_t memAddSize,
+                                      uint8_t maxWriteTime) {
     uint16_t memAddress16 = memAddress;
-    HAL_I2C_Mem_Write(&halI2C, addr << 1, memAddress16, memAddSize, bytes, size, DEFAULT_I2C_TIMEOUT);
-    HAL_Delay(maxWriteTime);
```

Comment:
This delay is required to give the signal time to complete without throwing errors. I think there should be a way to check flags to do this instead of just a flat delay, but I figured this would be good enough for now. You can either try to figure out a better way of doing it or add the HAL Delay back in.

Reply:
Ahh, I had thought the delay was related to the timeout. For waiting for the signal to propagate through, that is usually handled higher up because different systems may require different amounts of delay. (For example the EEPROM responds much faster over I2C then the BQ chip on the BMS). I'll certain re-run tests with the EEPROM without the delay written here to make sure everything is still ok.

Reply:
For now I reverted the change, but we should move any wait time to the EEPROM device since its specific to the EEPROM and not I2C.